File: src/BinarySearchTree.java
import java.util.*;

public class BinarySearchTree<T extends Comparable<T>> implements Iterable{
    class Node<T>{
        T data;
        Node<T> parent;
        Node<T> left;
        Node<T> right;

        Node(T data){
            this.data = data;
            this.parent = null;
            this.left = null;
            this.right = null;
        }

        public T getData(){
            return data;
        }

        @Override
        public String toString(){
            return data.toString();
        }
    }

    private Node<T> root;
    private String name;

    public String getName(){
        return this.name;
    }

    public void add(T value){
        root = add(null, root, value);
    }

    private Node<T> add(Node<T> parent, Node<T> root, T value){
        if(root == null) {
            root = new Node(value);
            root.parent = parent;
            return root;
        }
        if(value.compareTo(root.data) > 0)
            root.right = add(root, root.right, value);
        else if(value.compareTo(root.data) < 0)
            root.left = add(root, root.left, value);
        return root;
    }

    public void addAll(List<T> values){
        for(T value: values)
            add(value);
    }

    public BinarySearchTree(String name){
        this.name = name;
    }

    @Override
    public String toString(){
        StringBuilder string = new StringBuilder();
        preOrder(string, root);
        return "[" + name + "] " + string.toString();
    }

    public void preOrder(StringBuilder str, Node<T> node){
        if(node == null)
            return;

        str.append(node.data);

        if(node.left != null)
            str.append(" L:(");
        preOrder(str, node.left);

        if(node.right != null)
            str.append(" R:(");
        preOrder(str, node.right);

        if(node.left == null || node.right == null) {
            str.append(")");
        }
        else if(node.parent != null){
            if(node.parent.left != null || node.parent.right != null)
                str.append(")");
        }
    }

    @Override
    public Iterator<T> iterator() {
        Iterator<T> iterator = new BSTIterator(root);
        return iterator;
    }

    private class BSTIterator<T> implements Iterator<T>{
        private int index = 0;
        private Queue<Node<T>> nodes = new LinkedList();
        
        public BSTIterator(Node<T> root){
            inOrder(root, nodes);
        }

        private void inOrder(Node<T> root, Queue<Node<T>> queue){
            if(root == null)
                return;

            inOrder(root.left, queue);
            queue.add(root);
            inOrder(root.right, queue);
        }

        @Override
        public boolean hasNext() {
            return !nodes.isEmpty();
        }

        @Override
        public T next() {
            return nodes.poll().data;
        }

        @Override
        public void remove() {
            nodes.remove();
        }
    }

    public static class MergeRunnable<T extends Comparable<T>> implements Runnable{

        private BinarySearchTree<T> tree;
        private BinarySearchTree<T> merge;

        public MergeRunnable(BinarySearchTree<T> tree, BinarySearchTree<T> merge){
            this.tree = tree;
            this.merge = merge;
        }

        public void run(){
            this.tree.forEach(node -> merge.add((T) node));
        }
    }

    public static <T extends Comparable<T>> List<T> merge(List<BinarySearchTree<T>> bstlist){
        ArrayList<Thread> threads = new ArrayList<>();
        BinarySearchTree<T> resultTree = new BinarySearchTree("Merged Tree");
        ArrayList<T> result = new ArrayList<>();

        bstlist.forEach(bst -> threads.add(new Thread(new MergeRunnable(bst, resultTree))));
        threads.forEach(thread -> thread.start());
        //Ensure it finishes
        try{threads.get(threads.size()-1).join();}catch(InterruptedException ie){}
        resultTree.forEach(val -> result.add((T) val));
        return result;
    }

}

File: src/binarysearchtree.py
from io import StringIO

class Node:
    left = None
    right = None
    data = None

    def __init__(self):
        self.left = None
        self.right = None

    def set_data(self, data):
        self.data = data

    def __str__(self):
        return str(self.data)

    def __iter__(self):
        yield self.data

class BinarySearchTree:
    root = None
    name = None

    def __init__(self, name, root):
        self.root = root
        self.name = name

    def add(self,node,data):
        if node.data is None:
            node.set_data(data)
        else:
            if data < node.data:
                if node.left is not None:
                    self.add(node.left, data)
                else:
                    node.left = Node()
                    node.left.set_data(data)
            else:
                if node.right is not None:
                    self.add(node.right, data)
                else:
                    node.right = Node()
                    node.right.set_data(data)

    def add_all(self, *values):
        for value in values:
            self.add(self.root, value)
        return

    def inOrder(self,lst: list, root: Node):
        if(root == None):
            return []
        self.inOrder(lst, root.left);
        lst.append(root.data);
        self.inOrder(lst, root.right);

    def __iter__(self):
        list_of_nodes = []
        self.inOrder(list_of_nodes, self.root)
        for node in list_of_nodes:
            yield node

    def preOrder(self, string: StringIO, root: Node):

        if root is None:
            string.write(")")
            return

        string.write(str(root.data))
        if root.left is not None:
            string.write(" L:(")
        self.preOrder(string, root.left)
        if root.right is not None:
            string.write(" R:(")
        self.preOrder(string, root.right)

    def __str__(self):
        bst_string = StringIO()
        self.preOrder(string = bst_string, root = self.root)
        return bst_string.getvalue()

if __name__ == "__main__":
    tree = BinarySearchTree(name="Oak", root=Node())
    tree.add_all(5, 3, 9, 0) # adds the elements in the order 5, 3, 9, and then 0
    print(tree)

    t1 = BinarySearchTree(name="Oak", root=Node())
    t2 = BinarySearchTree(name="Birch", root=Node())
    t1.add_all(5, 3, 9, 0)
    t2.add_all(1, 0, 10, 2, 7)
    for x in t1:
        print(x)
    for x in t2:
        print(x)
